### PR TITLE
Add ELF program and section headers to core dump.

### PR DESCRIPTION
**Pull request by mikem8361**
Issue #12341

This allows the build id for the modules be extracted properly.

Add .eh_frame, .eh_frame_hdr and some other important sections to core dump.

Add managed module names to core dump so the symbol acquisition
tool can get symbol store keys for them.

Add the non memory/file backed memory regions to core dump so debuggers
like lldb can check if an address is "code" or not. Unwinding doesn't
work when there isn't a module/.so file present.

### CODE REVIEW

File: src/debug/createdump/memoryregion.h [hunk 1]
```diff
@@ -36,23 +47,57 @@ struct MemoryRegion
...
+
+    const uint32_t Permissions() const { return m_flags & MEMORY_REGION_FLAG_PERMISSIONS_MASK; }
+    const uint32_t Flags() const { return m_flags; }
+    const bool IsBackedByMemory() const { return (m_flags & MEMORY_REGION_FLAG_MEMORY_BACKED) != 0; }
     const uint64_t StartAddress() const { return m_startAddress; }
     const uint64_t EndAddress() const { return m_endAddress; }
     const uint64_t Size() const { return m_endAddress - m_startAddress; }
     const uint64_t Offset() const { return m_offset; }
-    const char* FileName() const { return m_fileName; }
+    char* FileName() const { return m_fileName; }
```
Comment (janvorli) [inline src/debug/createdump/memoryregion.h]: I wonder why was this change needed. There doesn't seem to be any newly added call to this method in the current PR and I assume the original version with the `const char*` return was ok before.

File: src/debug/createdump/crashinfo.cpp [hunk 2]
```diff
@@ -454,18 +488,134 @@ CrashInfo::GetDSOInfo()
...
+            return false;
+        }
         linkMapAddr = map.l_next;
     }
 
     return true;
 }
 
+inline bool
+NameCompare(char* name, const char* sectionName)
```
Comment (janvorli) [inline src/debug/createdump/crashinfo.cpp]: A nit - the first parameter could be `const`

**Comment by mikem8361** [comment]
I undid the change and made all references to the filename “const char *” (constructors, other functions, etc.).
